notebooks/query-dependent link analysis.ipynb

```diff
@@ -0,0 +1,1064 @@
+   "execution_count": null,
+   "id": "f9753fd2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "GOOGLE_APPLICATION_CREDENTIALS = os.getenv('GOOGLE_APPLICATION_CREDENTIALS')\n",
+    "\n",
+    "# Connect to service account\n",
+    "scope = ['https://spreadsheets.google.com/feeds'] \n",
+    "credentials = ServiceAccountCredentials.from_json_keyfile_name(GOOGLE_APPLICATION_CREDENTIALS, scope) \n",
+    "gc = gspread.authorize(credentials)\n",
+    "\n",
+    "# Import the data from google sheets\n",
+    "spreadsheet_key = '1x3lHpUkIm-KTnWrRJU11An1J2Yb4KJT8z0QpuXOm3e0' \n",
+    "book = gc.open_by_key(spreadsheet_key) \n",
+    "worksheet = book.worksheet('sab_pages') \n",
+    "table = worksheet.get_all_values()\n",
+    "\n",
+    "# Convert table data into a dataframe then set \n",
+    "df = pd.DataFrame(table[1:], columns=table[0])\n",
+    "sab_pages=set(df['pagePath'])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f865310f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Query neo4j\n",
+    "parameters = {}\n",
+    "parameters['pages'] = list(sab_pages)\n",
+    "\n",
+    "query = r\"\"\"\n",
+    "MATCH (c1:Cid)-[r:HYPERLINKS_TO|:USER_MOVEMENT]-(c2:Cid)\n",
+    "WHERE c1.name IN $pages\n",
+    "RETURN *\n",
+    "\"\"\"\n",
+    "g = getSubgraph(query, parameters)\n",
+    "showGraph(g)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "79e3b4a0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create a list of the nodes\n",
+    "nodes = list(g.nodes(data=True))\n",
+    "\n",
+    "# Only keep nodes that are also in sab_pages\n",
+    "sabNodes = [node for node in nodes if node[1]['properties']['name'] in sab_pages]\n",
+    "g = g.subgraph([node[0] for node in sabNodes])\n",
+    "\n",
+    "# Draw graph\n",
+    "showGraph(g)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4c35a7ff",
+   "metadata": {},
+   "source": [
+    "## Exploring the Start a Business subgraph"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "99b3668d",
+   "metadata": {},
+   "source": [
+    "Summary: \n",
+    "- Subgraph components = 274, 16, 3, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1\n",
+    "- Max shortest path length = 17\n",
+    "- Centrality metrics provide an estimate of node importance, pages such as \n",
+    "  '/get-information-about-a-company' and '/running-a-limited-company'\n",
+    "- Nodes with user movement between = 186\n",
+    "- User movement weight ranges from 6 to 34837"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "90df4bb9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Which nodes are in the list but not in the graph? \n",
+    "## Answer: two pages in the list no longer exist, so are not in the graph\n",
+    "\n",
+    "# List of node property 'name'\n",
+    "nodes_graph = []\n",
+    "for k,v in g.nodes(data=True):\n",
```

Comment:
Try list comprehension:

nodes_graph = [v['properties']['name'] for _,v in g.nodes(data=True)]

Reply:
Thank you! Refactored in commit cf48cce